<center>
    <h1>Il computer impara a scrivere</h1>
    <img src="robot1.jpg" style="width: 50%;" />
</center>


## Il nostro obiettivo:



<img src="robot4.jpg" style="width:20%;float:right;"/>
<img src="robot3.jpg" style="width:25%;float:right;"/>

*  costruire un generatore di titoli di giornale
*  vedere all'opera tre concetti fondamentali per la generazione del testo
    - Scelta casuale
    - Frequenza d'uso
    - Contesto



Per imparare a generare del testo ci servono degli **esempi**:
* usiamo un file con 50'000 titoli di giornali riguardanti la politica italiana (2018-2021)
* memorizziamo ogni titolo come la **sequenza di parole e punteggiatura** di cui è composto

In [1]:
import brightwrite
import importlib
importlib.reload(brightwrite)

# file con titoli di giornali riguardanti la politica italiana
filename = "data/politica.txt"

esempi = brightwrite.carica_esempi(filename)

brightwrite.stampa_esempi(esempi, 10)

  0%|          | 0/49717 [00:00<?, ?it/s]

Ho caricato 49717 esempi

1	Delrio : ' Mi sono occupato di Banca Etruria , era mio dovere ' .
2	Mdp : Speranza , superare renzismo .
3	Legittima difesa : Costa , no retromarcia .
4	Cassimatis , Fidatevi di me non valido .
5	SI : Vendola assolto da diffamazione Fitto .
6	Ala apre a Mattarellum , non escluderlo .
7	Cassimatis , è Pasqua , Grillo si redima .
8	Siria : Salvini , un regalo all ' Isis .
9	Mons . Fabio Dal Cin Arcivescovo Loreto .
10	Sciolti 4 comuni per infiltrazioni mafia .
11	Ap , guardate Borsa su ipotesi voto .


Costruiamo il **vocabolario** del nostro linguaggio:
* l'elenco di tutte le parole, e simboli, che compaiono almeno una volta negli esempi

In [2]:
vocabolario = brightwrite.costruisci_vocabolario(esempi)

  0%|          | 0/49717 [00:00<?, ?it/s]

Il vocabolario contiene 32399 elementi


In [3]:
# stampo tutti gli elementi del vocabolario
for elemento in vocabolario:
    print(f'{elemento:15}',end='\t')

%              	'              	+              	,              	.              	0              	00             	000            	004            	007            	01             	02             	023            	03             	04             	06             	062            	07             	073            	075            	08             	081            	09             	1              	10             	100            	1000           	101            	102            	103            	104            	106            	107            	108            	109            	11             	110            	111            	112            	1130           	114            	115            	116            	117            	1176           	118            	119            	12             	120            	1200           	123            	124            	125            	126            	128            	129            	13             	130            	1300           	131            	132            	134            	135     

<img src="scelta_uniforme.png" style="display:inline-block; float:right; width:50%" />     
<img src="dart.png" style="display:inline-block; float:right;margin-top:10%; width:20%" />

Possiamo generare un testo a partire dal vocabolario?

Sì, **scegliendo parole a caso!**

</div>

In [4]:
import random

lunghezza = 15
print(' '.join(random.choices(vocabolario, k=lunghezza)))

Riunisco stati Parenti programmatiche agitate Impediamo lista Stefani Filiera scelta tirate scarpe Calabria fattore Intervengano


Dove sono gli articoli, le preposizioni, i pronomi, virgole e punti?


<img src="scelta_pesata.png" style="display:inline-block; float:right;width:64%" /> 
<img src="dart.png" style="display:inline-block; float:right;margin-top:10%; width:20%" />

Scegliamo parole a caso **in proporzione alla frequenza d'uso negli esempi**.

In [5]:
vocabolario, frequenze = brightwrite.conta_frequenze(esempi)

# stampo i 30 elementi più frequenti
for idx in range(30):
    print(f'{frequenze[idx]:5} {vocabolario[idx]:30}', end='\t')

60544 .                             	37254 ,                             	28055 :                             	11105 di                            	10331 il                            	10203 a                             	 9309 '                             	 8638 e                             	 7667 la                            	 6931 per                           	 6358 in                            	 5482 Salvini                       	 4574 è                             	 4569 non                           	 4272 Pd                            	 4205 Conte                         	 4076 al                            	 4075 i                             	 3864 l                             	 3837 con                           	 3754 Di                            	 3592 del                           	 3207 Maio                          	 3077 M5S                           	 3006 le                            	 2947 Il                            	 2944 governo                       	 

In [6]:
lunghezza = 15

# impostiamo una probabilità di scelta proporzionale alla frequenza negli esempi
print(' '.join(random.choices(vocabolario, k=lunghezza, weights=frequenze)))

cure Siamo ' fattore nostra espulsa Alitalia . : lo Fico forte eccellenti alla .


Le parole hanno le giuste proporzioni **ma non sono legate tra loro**.

Le parole hanno un **contesto d'uso**:

"il presidente del..."

* ✔️ consiglio, partito, comitato...
* ❌ _ristrutturata_,  _avrei_, _banana_...

Data la parola precedente (o anche più parole), la scelta casuale usa le frequenze delle parole che compaiono dopo quella parola negli esempi.

<center>
    <img src="scelta_contesto.png" style="width:80%;">
</center>

Calcoliamo un **modello del linguaggio** che memorizza le frequenze per ogni contesto (da zero a quattro parole) osservato negli esempi:

In [7]:
lunghezza_massima_contesti = 4
modello = brightwrite.conta_frequenze_nei_contesti(esempi,lunghezza_massima_contesti)

  0%|          | 0/49717 [00:00<?, ?it/s]

Numero totale di parametri nel modello: 664675


In [8]:
brightwrite.mostra_frequenze_per_contesto(modello, '')

60544 .                              49717                                37254 ,                              28055 :                              11105 di                             10331 il                             10203 a                               9309 '                               8638 e                               7667 la                               ... (+17427)


In [15]:
brightwrite.mostra_frequenze_per_contesto(modello, 'presidente')

   73 .                                 72 della                             34 del                               24 dell                              24 Mattarella                         9 :                                  8 di                                 7 ,                                  5 è                                  5 con                              ... (+123)


In [12]:
brightwrite.mostra_frequenze_per_contesto(modello, 'presidente del')

   13 Consiglio                          7 Senato                             2 partito                            2 consiglio                          2 Piemonte                           1 Milan                              1 Pe                                 1 Anm                                1 Museo                              1 Parlamento                       ... (+3)


Generiamo il testo usando frequenze e contesti:

In [27]:
print(brightwrite.genera_testo(modello, lunghezza_max=15, interrompi_al_punto=True))

 Cento candidati dei 5 Stelle . Testa a testa tra centrosinistra anche nell industria italiana


In [30]:
print(brightwrite.genera_testo(modello, contesto='scontro', lunghezza_max=15, interrompi_al_punto=True))

scontro in diretta dalla fronda per Savona , io ho la testa del Nord a Conte


# ...e ChatGPT?

I concetti alla base di ChatGPT (e simili) **sono gli stessi:**

* osservare **tanti** *esempi* 
* per definire un *modello di probabilità* **complesso**
* che usa un **grande** *contesto*

<table>
<thead>
  <tr>
    <th></th>
    <th>BrightWrite</th>
    <th>GPT3</th>
    <th>Incremento</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>Quantità di esempi</td>
    <td>3.2 MB</td>
    <td>570 GB</td>
    <td>x 182400</td>
  </tr>
  <tr>
    <td>Numero parametri<br>nel modello</td>
    <td>664675</td>
    <td>175 B<br></td>
    <td>x 263000</td>
  </tr>
  <tr>
    <td>Dimensione contesto</td>
    <td>4</td>
    <td>2048</td>
    <td>x 512</td>
  </tr>
  <tr>
    <td>Tempo di apprendimento</td>
    <td> 6 s</td>
    <td>34 giorni</td>
    <td>-</td>
  </tr>
  <tr>
    <td>Hardware</td>
    <td>1 CPU</td>
    <td>1024 GPU</td>
    <td>-</td>
  </tr>
  <tr>
    <td>Costo</td>
    <td><0.01 \$</td>
    <td>4.6 M\$</td>
    <td>-</td>
  </tr>
</tbody>
</table>

<img src="robot2.png" style="margin:2%;width:35%;float:right;"/>
Le grandi dimensioni dei modelli GPT fanno emergere interessanti capacità di elaborazione dell'informazione.

Questi modelli raggiungeranno intelligenza paragonabile a quella umana? Il dibattito è aperto...

...nel frattempo, quando usiamo uno **strumento** GPT è sempre meglio **usare la nostra intelligenza e conoscenza** verificare la validità delle risposte ottenute.

<img src="qrcode.png" style="width:20%;float:left;"/>

Codice open source: 

https://github.com/aesuli/brightwrite